File: server/src/main/java/gg/mineral/server/entity/EntityImpl.java
package gg.mineral.server.entity;

import java.util.Random;
import java.util.concurrent.Callable;

import gg.mineral.api.entity.Entity;
import gg.mineral.api.math.MathUtil;
import gg.mineral.server.MinecraftServerImpl;
import gg.mineral.server.command.impl.KnockbackCommand;
import gg.mineral.server.entity.living.HumanImpl;
import gg.mineral.server.entity.living.human.PlayerImpl;
import gg.mineral.server.network.packet.play.clientbound.EntityStatusPacket;
import gg.mineral.server.network.packet.play.clientbound.EntityVelocityPacket;
import gg.mineral.server.network.packet.play.clientbound.SoundEffectPacket;
import gg.mineral.server.world.WorldImpl;
import lombok.Getter;
import lombok.Setter;
import lombok.val;

@Getter
public abstract class EntityImpl implements Callable<EntityImpl>, Entity, MathUtil {
    @Setter
    protected double x, y, z, motX, motY, motZ, headY;
    protected float yaw, pitch, lastYaw, lastPitch;
    @Setter
    protected boolean onGround = false;
    @Getter
    private final byte viewDistance = (byte) 10;
    protected final int id;
    @Getter
    private int currentTick;
    @Getter
    @Setter
    private boolean firstTick = true, firstAsyncTick = true, chunkUpdateNeeded = true;
    @Getter
    @Setter
    private int lastDamaged;
    @Getter
    private final Random random = new Random(); // TODO: better random implementation
    protected float width, height;
    @Getter
    protected WorldImpl world;
    protected final MinecraftServerImpl server;

    public EntityImpl(int id, WorldImpl world) {
        this.id = id;
        this.setWorld(world);
        this.server = world.getServer();
    }

    public void setWorld(WorldImpl world) {
        val oldWorld = getWorld();

        if (oldWorld != null)
            oldWorld.removeEntity(this.getId());
        world.addEntity(this);
        this.world = world;
    }

    public void setYaw(float yaw) {
        lastYaw = this.yaw;
        this.yaw = yaw;
    }

    public void setPitch(float pitch) {
        lastPitch = this.pitch;
        this.pitch = pitch;
    }

    public void tick() {
        currentTick++;

        if (motY < 0.005)
            motY = 0;

        if (motX < 0.005)
            motX = 0;

        if (motZ < 0.005)
            motZ = 0;

        if (motY > 0) {
            motY -= 0.08;
            motY *= 0.98;
        }

        if (motX > 0) {
            motX *= 0.91;
            if (onGround)
                motX *= 0.6;
        }

        if (motZ > 0) {
            motZ *= 0.91;
            if (onGround)
                motZ *= 0.6;
        }

    }

    public abstract void tickAsync();

    @Override
    public EntityImpl call() {
        tickAsync();
        return this;
    }

    @Override
    public void attack(int targetId) {
        val entity = server.getEntities().get(targetId);

        if (entity == null || entity.getCurrentTick() - entity.getLastDamaged() < 10)
            return;

        entity.setLastDamaged(entity.getCurrentTick());
        val statusPacket = new EntityStatusPacket(entity.getId(), (byte) 2);

        double motX = entity.getMotX();
        double motY = entity.getMotY();
        double motZ = entity.getMotZ();
        double x = KnockbackCommand.x;
        double y = KnockbackCommand.y;
        double z = KnockbackCommand.z;

        double extraX = KnockbackCommand.extraX;
        double extraY = KnockbackCommand.extraY;
        double extraZ = KnockbackCommand.extraZ;

        double yLimit = KnockbackCommand.yLimit;

        double friction = KnockbackCommand.friction;

        if (friction > 0) {
            motX /= friction;
            motY /= friction;
            motZ /= friction;
        } else {
            motX = 0;
            motY = 0;
            motZ = 0;
        }

        double distanceX = this.getX() - entity.getX();
        double distanceZ = this.getZ() - entity.getZ();

        double magnitude = Math.sqrt(distanceX * distanceX + distanceZ * distanceZ);

        motX -= distanceX / magnitude * x;
        motY += y;
        motZ -= distanceZ / magnitude * z;

        if (motY > yLimit)
            motY = yLimit;

        if (this instanceof HumanImpl human) {
            if (human.isExtraKnockback()) {
                float angle = (float) Math.toRadians(this.getYaw());
                double sin = -Math.sin(angle);
                double cos = Math.cos(angle);
                motX += extraX * sin;
                motY += extraY;
                motZ += extraZ * cos;
                this.setMotX(this.getMotX() * 0.6);
                this.setMotZ(this.getMotZ() * 0.6);
                human.setExtraKnockback(false);
            }
        }
        val velocityPacket = new EntityVelocityPacket(targetId, toVelocityUnits(motX),
                toVelocityUnits(motY), toVelocityUnits(motZ));

        if (entity instanceof PlayerImpl player)
            player.getConnection().queuePacket(statusPacket, velocityPacket);

        if (this instanceof PlayerImpl player)
            player.getConnection().queuePacket(statusPacket, new SoundEffectPacket("game.player.hurt",
                    toSoundUnits(entity.getX()), toSoundUnits(entity.getY()),
                    toSoundUnits(entity.getZ()), 1.0f, toPitchUnits((this.getRandom().nextFloat()
                            - this.getRandom().nextFloat()) * 0.2F + 1.0F)));
        entity.setMotX(motX);
        entity.setMotY(motY);
        entity.setMotZ(motZ);
    }
}

File: server/src/main/kotlin/gg/mineral/server/entity/EntityImpl.kt
package gg.mineral.server.entity

import gg.mineral.api.entity.Entity
import gg.mineral.api.math.MathUtil.cos
import gg.mineral.api.math.MathUtil.floor
import gg.mineral.api.math.MathUtil.sin
import gg.mineral.api.math.MathUtil.toPitchUnits
import gg.mineral.api.math.MathUtil.toRadians
import gg.mineral.api.math.MathUtil.toSoundUnits
import gg.mineral.api.math.MathUtil.toVelocityUnits
import gg.mineral.server.MinecraftServerImpl
import gg.mineral.server.command.impl.KnockbackCommand
import gg.mineral.server.entity.living.HumanImpl
import gg.mineral.server.entity.living.human.PlayerImpl
import gg.mineral.server.network.packet.play.clientbound.EntityStatusPacket
import gg.mineral.server.network.packet.play.clientbound.EntityVelocityPacket
import gg.mineral.server.network.packet.play.clientbound.SoundEffectPacket
import gg.mineral.server.world.WorldImpl
import gg.mineral.server.world.chunk.ChunkImpl
import gg.mineral.server.world.chunk.ChunkImpl.Companion.toKey
import java.util.*
import java.util.concurrent.ThreadLocalRandom
import kotlin.math.sqrt

abstract class EntityImpl(
    override val id: Int,
    world: WorldImpl
) : Entity {

    override val server: MinecraftServerImpl
        get() = world.server

    override var world = world
        set(value) {
            if (field === world) return

            field.removeEntity(this@EntityImpl.id)
            world.addEntity(this@EntityImpl)
            field = value

            if (this is PlayerImpl) this.visibleChunks.clear()
        }

    override var x: Double = 0.0
        set(value) {
            val oldChunkKey = toKey(field.toInt() shr 4, field.toInt() shr 4)
            val newChunkKey = toKey(value.toInt() shr 4, value.toInt() shr 4)
            world.updateEntityChunks(oldChunkKey, newChunkKey, this)
            field = value
        }
    override var y: Double = 70.0
    override var z: Double = 0.0
        set(value) {
            val oldChunkKey = toKey(field.toInt() shr 4, field.toInt() shr 4)
            val newChunkKey = toKey(value.toInt() shr 4, value.toInt() shr 4)
            world.updateEntityChunks(oldChunkKey, newChunkKey, this)
            field = value
        }
    override var motX: Double = 0.0
    override var motY: Double = 0.0
    override var motZ: Double = 0.0
    override var headY: Double = 0.0
    override var yaw: Float = 0f
        set(value) {
            lastYaw = field
            field = value
        }

    override var pitch: Float = 0f
        set(value) {
            lastPitch = field
            field = value
        }

    private var lastYaw: Float = 0f
    private var lastPitch: Float = 0f
    override var onGround: Boolean = false
    override var viewDistance = 10.toByte()
    override var currentTick = 0
    override var currentAsyncTick = 0

    var firstTick = true
    var chunkUpdateNeeded = true
    private var lastDamaged = 0

    protected var width: Float = 0f
    protected var height: Float = 0f

    private val random: Random
        get() = ThreadLocalRandom.current()

    open fun tick() {
        currentTick++

        world.updatePosition(this)
    }

    init {
        world.addEntity(this)
    }

    override fun attack(targetId: Int) {
        val entity = world.getEntity(targetId) ?: return

        if (entity.currentTick - entity.lastDamaged < 10) return

        entity.lastDamaged = entity.currentTick
        val statusPacket = EntityStatusPacket(entity.id, 2.toByte())

        var motX: Double = entity.motX
        var motY: Double = entity.motY
        var motZ: Double = entity.motZ
        val x = KnockbackCommand.x
        val y = KnockbackCommand.y
        val z = KnockbackCommand.z

        val extraX = KnockbackCommand.extraX
        val extraY = KnockbackCommand.extraY
        val extraZ = KnockbackCommand.extraZ

        val yLimit = KnockbackCommand.yLimit

        val friction = KnockbackCommand.friction

        if (friction > 0) {
            motX /= friction
            motY /= friction
            motZ /= friction
        } else {
            motX = 0.0
            motY = 0.0
            motZ = 0.0
        }

        val distanceX = this.x - entity.x
        val distanceZ = this.z - entity.z

        val magnitude = sqrt(distanceX * distanceX + distanceZ * distanceZ)

        motX -= distanceX / magnitude * x
        motY += y
        motZ -= distanceZ / magnitude * z

        if (motY > yLimit) motY = yLimit

        if (this is HumanImpl) {
            if (this.extraKnockback) {
                val angle = toRadians(this.yaw)
                val sin = -sin(angle.toDouble())
                val cos = cos(angle.toDouble())
                motX += extraX * sin
                motY += extraY
                motZ += extraZ * cos
                this.motX *= 0.6
                this.motZ *= 0.6
                this.extraKnockback = false
            }
        }
        val velocityPacket = EntityVelocityPacket(
            targetId, toVelocityUnits(motX),
            toVelocityUnits(motY), toVelocityUnits(motZ)
        )

        if (entity is PlayerImpl) entity.connection.queuePacket(velocityPacket)
        if (this is PlayerImpl) {
            val chunkX = floor(this.x / 16).toByte()
            val chunkZ = floor(this.z / 16).toByte()

            for (xOffset in -2..2) {
                val cX = (chunkX + xOffset).toByte()
                for (zOffset in -2..2) {
                    val cZ = (chunkZ + zOffset).toByte()
                    val chunkKey = toKey(cX, cZ)
                    val chunk = world.getChunk(chunkKey) as ChunkImpl
                    val entityIterator = chunk.entities.iterator()
                    while (entityIterator.hasNext()) {
                        val entityId = entityIterator.nextInt()
                        val player = world.getEntity(entityId) as? PlayerImpl ?: continue
                        if (!player.visibleEntities.containsKey(entity.id) && player.id != entityId) continue

                        player.connection.queuePacket(statusPacket)

                        if (player.id != entity.id) {
                            player.connection.queuePacket(
                                SoundEffectPacket(
                                    "game.player.hurt",
                                    toSoundUnits(entity.x), toSoundUnits(entity.y),
                                    toSoundUnits(entity.z), 1.0f, toPitchUnits(
                                        ((random.nextFloat()
                                                - random.nextFloat()) * 0.2f + 1.0f).toDouble()
                                    )
                                )
                            )
                        }
                    }
                }
            }
        }

        entity.motX = motX
        entity.motY = motY
        entity.motZ = motZ
    }

    fun cleanup() {
        world.removeEntity(this.id)
    }
}
